streams/hash/others/hash_functions/gost/gost.cpp

```diff
@@ -0,0 +1,879 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  Use this program  at  your own risk!
+ * https://github.com/rhash/RHash/blob/master/librhash/gost.c
+ */
+
+#include <cstring>
+#include "../whirlpool/byte_order.h"
+#include "gost.h"
+
+extern unsigned rhash_gost_sbox[4][256];
+extern unsigned rhash_gost_sbox_cryptpro[4][256];
+
+/**
+ * Initialize algorithm context before calculaing hash
+ * with test parameters set.
+ *
+ * @param ctx context to initialize
+ */
+void rhash_gost_init(gost_ctx *ctx)
+{
+    memset(ctx, 0, sizeof(gost_ctx));
+}
+
+/**
+ * Initialize GOST algorithm context with CryptoPro parameter set.
+ *
+ * @param ctx context to initialize
+ */
+void rhash_gost_cryptopro_init(gost_ctx *ctx)
+{
+    rhash_gost_init(ctx);
+    ctx->cryptpro = 1;
+}
+
+#if defined(__GNUC__) && defined(CPU_IA32) && !defined(__clang__) && !defined(RHASH_NO_ASM)
+# define USE_GCC_ASM_IA32
+#elif defined(__GNUC__) && defined(CPU_X64) && !defined(RHASH_NO_ASM)
+# define USE_GCC_ASM_X64
+#endif
+
+/*
+ *  A macro that performs a full encryption round of GOST 28147-89.
+ *  Temporary variables tmp assumed and variables r and l for left and right
+ *  blocks.
+ */
+# define GOST_ENCRYPT_ROUND(key1, key2, sbox) \
+	tmp = (key1) + r; \
+	l ^= (sbox)[tmp & 0xff] ^ ((sbox) + 256)[(tmp >> 8) & 0xff] ^ \
+		((sbox) + 512)[(tmp >> 16) & 0xff] ^ ((sbox) + 768)[tmp >> 24]; \
+	tmp = (key2) + l; \
+	r ^= (sbox)[tmp & 0xff] ^ ((sbox) + 256)[(tmp >> 8) & 0xff] ^ \
+		((sbox) + 512)[(tmp >> 16) & 0xff] ^ ((sbox) + 768)[tmp >> 24];
+
+# define GOST_ENCRYPT_ROUND_L(key1, key2, sbox) \
+	tmp = (key1) + r; \
+	l ^= (sbox)[tmp & 0xff] ^ ((sbox) + 256)[(tmp >> 8) & 0xff] ^ \
+		((sbox) + 512)[(tmp >> 16) & 0xff] ^ ((sbox) + 768)[tmp >> 24]; \
+
+# define GOST_ENCRYPT_ROUND_R(key1, key2, sbox) \
+	tmp = (key2) + l; \
+	r ^= (sbox)[tmp & 0xff] ^ ((sbox) + 256)[(tmp >> 8) & 0xff] ^ \
+		((sbox) + 512)[(tmp >> 16) & 0xff] ^ ((sbox) + 768)[tmp >> 24];
+
+/* encrypt a block with the given key */
+/* ph4r05: round reduced version from the macro below */
+void gost_encrypt_fnc(unsigned int * result, unsigned int i, const unsigned int * key, const unsigned int * hash,
+                      const unsigned int * sbox, unsigned nr)
+{
+    unsigned l, r, tmp, jj;
+    r = hash[i], l = hash[i + 1];
+    for(jj = 0; jj < 12 && 2*jj < nr; ++jj){
+        GOST_ENCRYPT_ROUND_L(key[(2 * jj + 0) % 8], key[(2 * jj + 1) % 8], sbox)
+        if (2*jj + 1 < nr) {
+            GOST_ENCRYPT_ROUND_R(key[(2 * jj + 0) % 8], key[(2 * jj + 1) % 8], sbox)
+        }
+    }
+    for(jj = 12; jj < 16 && 2*jj < nr; ++jj){
+        GOST_ENCRYPT_ROUND_L(key[2 * (16 - jj) - 1], key[2 * (16 - jj) - 2], sbox)
+        if (2*jj + 1 < nr) {
+            GOST_ENCRYPT_ROUND_R(key[2 * (16 - jj) - 1], key[2 * (16 - jj) - 2], sbox)
+        }
+    }
+
+    result[i] = l, result[i + 1] = r;
+}
+
+/* encrypt a block with the given key */
```

Comment:
Shouldn't we comment this out? For the first look, I thought it is your code (`nr`), but then I saw it is just reference of the former code, which is substituted by the function call.

Reply:
This is original piece of code. I think it is not used by anything so it should be optimized out. I wanted to keep it for the reference. Comment out - like to avoid some confusion?